hw/ip/otbn/dv/rig/rig/gens/bad_branch.py

```diff
@@ -25,81 +20,14 @@ class BadBranch(SnippetGen):
 
     ends_program = True
 
-    def __init__(self, cfg: Config, insns_file: InsnsFile) -> None:
-        super().__init__()
-
-        self.insns = []
-        self.weights = []
-
-        self.beq = self._get_named_insn(insns_file, 'beq')
-        self.bne = self._get_named_insn(insns_file, 'bne')
-
-        # beq and bne expect operands: grs1, grs2, offset
-        for insn in [self.beq, self.bne]:
-            if not (len(insn.operands) == 3 and
-                    isinstance(insn.operands[0].op_type, RegOperandType) and
-                    insn.operands[0].op_type.reg_type == 'gpr' and
-                    not insn.operands[0].op_type.is_dest() and
-                    isinstance(insn.operands[1].op_type, RegOperandType) and
-                    insn.operands[1].op_type.reg_type == 'gpr' and
-                    not insn.operands[1].op_type.is_dest() and
-                    isinstance(insn.operands[2].op_type, ImmOperandType) and
-                    insn.operands[2].op_type.signed):
-                raise RuntimeError('{} instruction from instructions file is not '
-                                   'the shape expected by the Branch generator.'
-                                   .format(insn.mnemonic))
-
-        self.imm_op_type = self.bne.operands[2].op_type
-
-        for insn in insns_file.insns:
-            if insn.mnemonic in ['beq', 'bne']:
-                weight = cfg.insn_weights.get(insn.mnemonic)
-                if weight > 0:
-                    self.weights.append(weight)
-                    self.insns.append(insn)
-
-        # Check that at least one instruction has a positive weight.
-        assert len(self.insns) == len(self.weights)
-        if not self.weights:
-            self.disabled = True
-
-    def gen(self,
-            cont: GenCont,
-            model: Model,
-            program: Program) -> Optional[GenRet]:
-
-        # Pick maximum and minimum address range for the current PC.
-        imm_rng = self.imm_op_type.get_op_val_range(model.pc)
-        assert imm_rng is not None
-
-        min_addr, max_addr = imm_rng
-
-        # Get known registers. We always have x0.
-        # So it should never fail.
-        known_regs = model.regs_with_known_vals('gpr')
-        assert known_regs is not None
-
-        # Pick a random register among known registers.
-        idx, value = random.choice(known_regs)
-
-        equals = []
-        not_equals = []
-
-        for reg_idx, reg_val in known_regs:
-            if reg_val == value:
-                equals.append(reg_idx)
-            else:
-                not_equals.append(reg_idx)
-
-        # Get the chosen base register index as grs1 and grs2 operand. This is
-        # because we want to branch to the faulty addresses with this snippet.
-        op_val_grs1 = idx
-        assert op_val_grs1 is not None
-
+    def cal_offset(self,
```

Comment:
Nit: I think that most functions like this are called "pick_$foo". Would you mind renaming?